runtime/hashing/common/src/aws/smithy/kotlin/runtime/hashing/HashingSink.kt

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.hashing
+
+import aws.smithy.kotlin.runtime.io.SdkSink
+import aws.smithy.kotlin.runtime.io.internal.SdkSinkObserver
+
+/**
+ * A sink which hashes data as it is being written to
+ * @param hash The [HashFunction] to hash data with
+ * @param sink the [SdkSink] to hash
+ */
+public class HashingSink(public val hash: HashFunction, sink: SdkSink) : HashFunction by hash, SdkSinkObserver(sink) {
+    override fun observe(data: ByteArray, offset: Int, length: Int) {
+        update(data, offset, length)
+    }
+}
```

Comment:
**Correctness**: This largely duplicates [the `private` class in **Canonicalizer.kt**](https://github.com/awslabs/smithy-kotlin/blob/855a77a3ea1e9a4b7f2fe005e297de13c7559732/runtime/auth/aws-signing-default/common/src/aws/smithy/kotlin/runtime/auth/awssigning/Canonicalizer.kt#L182-L189). We should remove that and update the surrounding code to use this.